### PR TITLE
adds `humility qemu`

### PR DESCRIPTION
**Pull request by beezow**
Adds new `humility qemu` to start the app within qemu.

### CODE REVIEW

File: cmd/qemu/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    //parse port from args, this is the port gdb-server will listen on
+    let serv_config = format!("tcp::{}", subargs.port);
+
+    let work_dir = tempfile::tempdir()?;
+
+    // extract bin to run in qemu
+    hubris
+        .extract_file_to("img/final.bin", &work_dir.path().join("final.bin"))?;
+
+    // extract elf to pase the boot address
```
Comment (furquan-rivos) [inline cmd/qemu/src/lib.rs]: s/pase/parse

File: cmd/qemu/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    hubris
+        .extract_file_to("img/final.elf", &work_dir.path().join("final.elf"))?;
+
+    // extract qemu runner from hubris archhubris
```
Comment (furquan-rivos) [inline cmd/qemu/src/lib.rs]: s/archhubris/archive?

File: cmd/qemu/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    hubris
+        .extract_file_to("debug/qemu.sh", &work_dir.path().join("qemu.sh"))
+        .context("No 'qemu.sh' archive found in the hubris archive")?;
+    let qemu_cmd = fs::read_to_string(&work_dir.path().join("qemu.sh"))
+        .expect("Could not find 'qemu.sh' command");
+    //let qemu_cmd = format!(qemu_cmd.to_string() , boot_addr, serv_config);
```
Comment (furquan-rivos) [inline cmd/qemu/src/lib.rs]: drop commented out code?

File: README.md.in [hunk 4]
```diff
@@ -91,6 +91,9 @@ which can have the following values:
   a halt. To recover from this condition, send an explicit ^C to the
   running GDB and continue from the resulting stop.
 
+- `qemu`: Attach via Qemu's GDB server, which is presumed to have the interface available on localhost:1234
```
Comment (furquan-rivos) [inline README.md.in]: Isn't the default 3333 in the code?

File: README.md [hunk 5]
```diff
@@ -1216,7 +1220,10 @@ VSC8552; this is indicated with `--` in the relevant table positions.
 
 ### `humility openocd`
 
-This command launches OpenOCD based on the config file in a build archive
+This command launches OpenOCD based on the config file in a build archive,
+which then allows one to connect with either GDB or directly via telnet.
+If the intention is to only run GDB, note that `humility gdb --run-openocd`
+will both run OpenOCD and run a foreground GDB that is connected to it.
```
Comment (furquan-rivos) [inline README.md]: Not related change

File: cmd/qemu/src/lib.rs (hunk 1 above)
Comment (beezow) [reply cmd/qemu/src/lib.rs]: fixed

File: cmd/qemu/src/lib.rs (hunk 2 above)
Comment (beezow) [reply cmd/qemu/src/lib.rs]: fixed

File: cmd/qemu/src/lib.rs (hunk 3 above)
Comment (beezow) [reply cmd/qemu/src/lib.rs]: yep

File: README.md.in (hunk 4 above)
Comment (beezow) [reply README.md.in]: Good catch, I changed it throughout the development process.

File: README.md (hunk 5 above)
Comment (beezow) [reply README.md]: Like the other PRs, the README.md was out of date with the README.md.in.

**Comment by furquan-rivos** [comment]
Looks like something went wrong here? I see 21 commits instead of 3.

**Review by furquan-rivos — APPROVED** [review]

**Pull request merged**